Support extra inputs for subgraph ops

samskalicky opened the pull request:
## Description ##
Support additional inputs to custom subgraph ops that are not direct dependencies to ops in the subgraph. This will enable various use cases: custom control flow ops, custom ops that maintain a state that should be saved/loaded, etc.

Highlights:
- Added test that uses a graph pass (`addInputPass`) to add a new custom input to the subgraph op
- Added new optional argument (`clear`) to `hybridize` & `optimize_for` APIs in Gluon Block to enable multiple optimizations
- refactored lib_api.h JSON utilities
- added new Graph data structure utilities to simplify custom graph passes
- refactored custom op registration
- enhanced custom subgraph op to support additional inputs to subgraph op that is not an input to ops in the subgraph
- updated subgraph & graph pass READMEs
- Added error messaging from external library

### Extra Inputs for Subgraph Ops ###
This feature is enabled by combining the custom partitioner and custom graph pass features. We leverage the custom graph partitioner to create the subgraph op and perform regular partitioning. Then, leverage a custom graph pass to find the newly created subgraph ops and add extra inputs.

The "lib_subgraph" example in example/extensions/lib_subgraph has been extended to optionally add extra inputs to the subgraph op. It shows finding the subgraph nodes that were created during partitioning, and adding extra inputs. Heres a snippet (but I encourage you to go look at the code change in subgraph_lib.cc):
```
Graph *g = Graph::fromString(in_graph);
//find node with '_custom_subgraph_op' op type
for(Node* n : g->nodes) {
  if(n->op.compare("_custom_subgraph_op") == 0) {
    //set extra input
    n->attrs[MX_STR_EXTRA_INPUTS] = std::to_string(1);

    // ... create a new node in the graph, write it up to the subgraph node ...

    // add a corresponding tensor for this input
    new_node.alloc_arg({1},MXContext::CPU(0),kFloat32);
  }
}
```

Modifying the `Graph` object updates the symbol, and calling the `alloc_arg` creates a tensor that will be added to the params for the model. 

### Custom Subgraph Op Enhancements ###
Added new custom lambda functions for subgraph ops instead of using default `DefaultSubgraphOp*` from src/operator/subgraph/common.h. These functions "hide" the extra inputs to the subgraph op from processing for inferring shapes, types, storage types for the ops inside the subgraph. This is configured via the `MX_STR_EXTRA_INPUTS` attribute that can be set on subgraph ops.

### Gluon API Changes ###
Previously hybridize used to clear the `cachedOp` and `cached_graph` on every call. In this PR we add a new argument `clear` that defaults to `True` to preserve current behavior. When set to `False` it allows re-using the previous `cached_graph` to chain multiple optimization calls together. Heres an example:
```
net = nn.SymbolBlock(sym, inputs)
net.optimize_for(x, backend="opt1")
net.optimize_for(x, backend="opt2", clear=False)
```

### Graph Pass Enhancements ###
Raises the level of abstraction for custom graph pass writers by adding a `Graph` class to represent the symbol graph. Builds on top of the JSON support. Enables users to easily add/remove nodes in the graph, connect/disconnect nodes, modify attributes, and add new args/aux, etc. At a high level, this means modifying operators, fusing operators (ie. replacing Conv, BN, Act chains with a custom op), fusing params, etc. All conversion from JSON string to Graph object and back is now built-in. Users can just use the Graph object directly. Heres an example:
```
MXReturnValue addInputPass(Graph *g,
			   const std::unordered_map<std::string, std::string>& options) { 
  // ... do custom graph pass ...

  return MX_SUCCESS;
}
```
Heres an example adding a new node in the graph and then allocating a new arg param for it:
```
Node *n = g->addNode("myNode","null");
n->alloc_arg({1},MXContext::CPU(0),kFloat32);
```
Heres an example creating an edge between two nodes:
```
n1->outputs[0] = {n2,1};
n2->inputs[1] = {n1,0};
```
Here node `n1` produces an output at index 0 that is consumed by node `n2` on the input at index 1.
![example_nodes](https://user-images.githubusercontent.com/1541993/88966633-7aac6300-d261-11ea-80de-ae3490640ff7.png)

### JSON parsing enhancements ###
Removed the JsonParser class and refactored the parsing functions into the JsonVal class as static function. Before you used to do this to parse a string into a JSON object:
```
JsonParser parser;
JsonVal json_val = parser.parse_to_json(subgraph_sym);
```
And now you can just call the static function:
```
JsonVal json_val = JsonVal::parse(subgraph_sym);
```

### Test changes ###
Refactors `test_subgraph.py` and removes the old `Symbol.bind` uses now that they are deprecated in 2.0. 

### Error messaging ###
Added support to enable returning custom error messages from external library API calls. For example:

```
Traceback (most recent call last):
  File "test_subgraph.py", line 35, in <module>
    mx.library.load(path)
  File "/home/ubuntu/subgraph_inputs/python/mxnet/library.py", line 56, in load
    check_call(_LIB.MXLoadLib(chararr, mx_uint(verbose_val)))
  File "/home/ubuntu/subgraph_inputs/python/mxnet/base.py", line 246, in check_call
    raise get_last_ffi_error()
mxnet.base.MXNetError: Traceback (most recent call last):
  File "/home/ubuntu/subgraph_inputs/src/c_api/c_api.cc", line 1465
MXNetError: Library failed to initialize
Extension Traceback:
	subgraph_lib.cc[329]: MXNet version 20000 not supported by custom library
```
Notice the new *Extension Traceback* at the bottom. On the user-side, a new macro simplifies adding error messages:
```
MX_ERROR_MSG << "MXNet version " << version << " not supported by custom library" << std::endl;
```
Users can enqueue multiple error messages by using `MX_ERROR_MSG` multiple times. Error messages are printed when returning `MX_FAIL` from any extension API (ie. `initialize`, `supportedOps`, etc that return an `MXReturnValue`).

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

samskalicky commented:
@MoisesHer @Kh4L @HahTK for review

samskalicky commented:
@mseth10

samskalicky commented:
@eric-haibin-lin @MoisesHer what do you think about changing the graph pass API from taking a String (symbol JSON) to a Graph object like:
```
MXReturnValue addInputPass(Graph* g,
			   const std::unordered_map<std::string, std::string>& options,
			   const std::unordered_map<std::string, MXTensor>& args,
			   const std::unordered_map<std::string, MXTensor>& aux,
			   const PassResource& res)
```
And then you just modify the graph object directly, dont need to even consider the string or the JSON encoding?

samskalicky commented:
@mseth10 @rondogency should we change `supportedOps` to use the `Graph` class too instead of passing the JSON string like:
```
MXReturnValue mySupportedOps(const Graph& g,
                             std::vector<int>* ids,
                             const std::unordered_map<std::string, std::string>& options)
```

eric-haibin-lin commented on python/mxnet/gluon/block.py:
```diff
@@ -1225,6 +1226,7 @@ def hybridize(self, active=True, backend=None, backend_opts=None, **kwargs):
             The name of backend, as registered in `SubgraphBackendRegistry`, default None
         backend_opts : dict of user-specified options to pass to the backend for partitioning, optional
             Passed on to `PrePartition` and `PostPartition` functions of `SubgraphProperty`
+        clear : clears any previous optimizations
```
Is there a case where we want to call `hybridize(clear=False)`?

samskalicky replied:
Yes, Previously hybridize used to clear the cachedOp and cached_graph on every call. In this PR we add a new argument clear that defaults to True to preserve current behavior. When set to False it allows re-using the previous cached_graph to chain multiple optimization calls together. Heres an example:
```
net = nn.SymbolBlock(sym, inputs)
net.optimize_for(x, backend="opt1")
net.optimize_for(x, backend="opt2", clear=False)
```

samskalicky replied:
Maybe the 1st call to `optimize_for` is to partition the graph, and the 2nd call would be to run a custom graph pass for example.

eric-haibin-lin replied:
cool got it

MoisesHer commented:
> @eric-haibin-lin @MoisesHer what do you think about changing the graph pass API from taking a String (symbol JSON) to a Graph object like:
From a user perspective (someone defining a pass) I think a graph is preferred. I guess the user will not have to re-implement class Graph  anymore, and will skip conversion from String <-> graph

HahTK commented on example/extensions/lib_subgraph/subgraph_lib.cc:
```diff
@@ -331,6 +332,46 @@ REGISTER_PARTITIONER(mySelect)
+  return MX_SUCCESS;
+}
+
+REGISTER_PASS(addInputPass)
```
I propose, not overloading the optime_for() path to apply custom passes.
Instead we should just have an API to apply_pass() or apply_pass_for().

This makes the intent clear and unambiguous.

samskalicky replied:
Graph passes usage through optimize_for is already merged in #17885. We're not changing how graph passes work in this PR

samskalicky commented on example/extensions/lib_subgraph/subgraph_lib.cc:
```diff
@@ -331,6 +330,46 @@ REGISTER_PARTITIONER(mySelect)
 .setReviewSubgraph("strategy1", myReviewSubgraph);
 
+/* \brief a basic pass that adds a new input for subgraph ops */
+MXReturnValue addInputPass(const std::string& in_graph, const std::string** out_graph,
```
To add a new input you have to do 2 things:
1. Create the new node and add it to the graph:
```
Node* input = new Node();
input->name = "my_new_input";
input->op = "null";
g->nodes.push_back(input);
g->inputs.push_back(input);
```
2. Create a new tensor to correspond to the new node:
```
MXTensor* arg_ = res.alloc_arg("my_new_input",{1},MXContext::CPU(0),kFloat32);
```
Notice that the node name and tensor name must match. 

> The new input has unknown and arbitrary tensor shape.
> Do we handle that somewhere ?
> 

In step 2 above you specify the shape when creating the tensor. In the example its `{1}`.

> Also is this added to both the
> a) subgraph_op and
> b) the subgraph attr of the subgraph_op ?
> 
> We probably do not want that in the latter.
> 

After creating the node in the graph and the corresponding tensor, you'll probably want to connect that node to your subgraph node to do anything interesting with it like:
```
//connect new input to node
input->outputs.push_back({n,(int)(n->inputs.size())});
//connect node to new input
n->inputs.push_back({input,0});
```
Where `n` is your subgraph node. 

Also, you'll need to set the attr on your subgraph op to make sure this input doesnt participate in shape/type inference like: `n->attrs[MX_STR_EXTRA_INPUTS] = std::to_string(1);`

> Also how is shape handled ?
> The input tensor can be an take arbitrary and unknown shape at this point in the flow.
> Once known, this shape is unknown to user and needs to be remembered.
> Therefore, this op likely needs the shape attr with a byte dtype.
> Do these already work with this update ? or is it upto the user to add ?

The example above shows setting `kFloat32` but you can also set:
```
enum MXDType {
  kFloat32 = 0,
  kFloat64 = 1,
  kFloat16 = 2,
  kUint8 = 3,
  kInt32 = 4,
  kInt8  = 5,
  kInt64 = 6,
};
```

samskalicky commented:
Todo: https://github.com/apache/incubator-mxnet/pull/18815/files#diff-3ac18ad0381b998b34f8d31f6a566dc8

mseth10 commented:
> @mseth10 @rondogency should we change `supportedOps` to use the `Graph` class too instead of passing the JSON string like:
> 
> ```
> MXReturnValue mySupportedOps(const Graph& g,
>                              std::vector<int>* ids,
>                              const std::unordered_map<std::string, std::string>& options)
> ```

Would we need to provide users with a Graph parser then, just like we provide a JsonParser currently?

samskalicky commented:
@mseth10 just pushed some changes, take a look and let me know what you think. Now conversion from JSON string to Graph object happens in the callSupportedOps function, and the Graph object is passed to the supportedOps API directly. So users dont even need to know about JSON strings.

mseth10 commented:
> @mseth10 just pushed some changes, take a look and let me know what you think. Now conversion from JSON string to Graph object happens in the callSupportedOps function, and the Graph object is passed to the supportedOps API directly. So users dont even need to know about JSON strings.

@samskalicky looks good to me, it's definitely a better user experience.

samskalicky commented:
> > @eric-haibin-lin @MoisesHer what do you think about changing the graph pass API from taking a String (symbol JSON) to a Graph object like:
>
> From a user perspective (someone defining a pass) I think a graph is preferred. I guess the user will not have to re-implement class Graph  anymore, and will skip conversion from String <-> graph

Can you review again? updated Graph to embed the args/aux as tensors within the appropriate Node's. And made it so that you can just call `node->alloc_arg/aux()` to allocate a tensor for a node in the graph. Per our offline discussion this simplifies graph passes significantly.

samskalicky commented:
@eric-haibin-lin @MoisesHer @Kh4L @ptrendx @PatricZhao @ZhennanQin @mseth10 @rondogency All changes are now complete, docs are updated. Anxiously awaiting your review/feedback. Thanks!

MoisesHer commented:
Thanks Sam, I think the way to add new params is very clear
just couple of questions / minor things about documentation on lib_pass/README.md

- is it required the double connection between nodes?  Here:
```
n1->outputs.push_back({n2,1});
n2->inputs.push_back({n1,0});
```
will not be enough to do something like
` n2->inputs[1].node = n1`

- Can you add the same figure in README as you have here?
- Example call for adding params (alloc_arg) here has a different signature as in lib_pass/README.md (+- input_name)

Kh4L commented on include/mxnet/lib_api.h:
```diff
@@ -748,62 +759,410 @@ struct JsonParser {
+      n->op = node.map[JsonVal("op")].str;
+      n->name = node.map[JsonVal("name")].str;
+
+      // if op is null its an input to the graph
```
Nit: `it's`

samskalicky replied:
fixed!

samskalicky commented:
> Thanks Sam, I think the way to add new params is very clear
> just couple of questions / minor things about documentation on lib_pass/README.md
> 
> * is it required the double connection between nodes?  Here:
> 
> ```
> n1->outputs.push_back({n2,1});
> n2->inputs.push_back({n1,0});
> ```
> 
> will not be enough to do something like
> ` n2->inputs[1].node = n1`
> 
> * Can you add the same figure in README as you have here?
> * Example call for adding params (alloc_arg) here has a different signature as in lib_pass/README.md (+- input_name)

Thanks Moises, I updated the PR description to fix the alloc_arg discrepancy and added the figure to the README. Per our offline discussion, `Graph` class has bidirectional connections so users should create connection in both ways when they add new nodes.

mseth10 commented on include/mxnet/lib_api.h:
```diff
@@ -748,62 +759,410 @@ struct JsonParser {
+    return n;
+  }
+  /* \brief get node at index in graph */
+  Node* getNode(size_t idx) {
```
why do we need two `getNode` functions?

samskalicky replied:
one is const

szha commented:
great work!
